File: Cpp/Lexer/lexer.cpp
#include "lexer.hpp"

Lexer::Lexer(string input) {
    this->input = input;
    this->curr  = 0;
    this->peek  = 0;
    this->ch    = '\0';

    this->advance();
}

Token Lexer::nextToken() {
    while (isspace(this->ch))
        advance();

    Token tok;

    // if a peek is possible, try checking double tokens
    if (peek < input.length()) {
        string ds{input[curr], input[peek]};
        if (ds == "//") return Token(::COMMENT, readComment());
        if (DOUBLE_TOKEN_MAP.find(ds) != DOUBLE_TOKEN_MAP.end()) {
            advance();
            advance();
            return Token(DOUBLE_TOKEN_MAP.at(ds), ds);
        }
    }

    switch (this->ch) {
        case '\0': return Token(::_EOF, "\0"); break;
        case '!':  tok = Token(::EXCLAMATION, "!"); break;
        case '@':  tok = Token(::AT, "@"); break;
        case '#':  tok = Token(::HASHTAG, "#"); break;
        case '$':  tok = Token(::DOLLAR, "$"); break;
        case '%':  tok = Token(::PERCENT, "%"); break;
        case '^':  tok = Token(::CARET, "^"); break;
        case '&':  tok = Token(::AMPERSAND, "&"); break;
        case '*':  tok = Token(::ASTERISK, "*"); break;
        case '(':  tok = Token(::LPAREN, "("); break;
        case ')':  tok = Token(::RPAREN, ")"); break;
        case '-':  tok = Token(::MINUS, "-"); break;
        case '_':  tok = Token(::UNDERSCORE, "_"); break;
        case '+':  tok = Token(::PLUS, "+"); break;
        case '=':  tok = Token(::ASSIGN, "="); break;
        case '[':  tok = Token(::LBRACKET, "["); break;
        case ']':  tok = Token(::RBRACKET, "]"); break;
        case '{':  tok = Token(::LBRACE, "{"); break;
        case '}':  tok = Token(::RBRACE, "}"); break;
        case ';':  tok = Token(::SEMICOLON, ";"); break;
        case ':':  tok = Token(::COLON, ":"); break;
        case '\'': tok = Token(::CHAR, readChar()); break;
        case '"':  tok = Token(::STRING, readString()); break;
        case ',':  tok = Token(::COMMA, ","); break;
        case '.':  tok = Token(::PERIOD, "."); break;
        case '<':  tok = Token(::LESSTHAN, "<"); break;
        case '>':  tok = Token(::GREATERTHAN, ">"); break;
        case '/':  tok = Token(::SLASH, "/"); break;
        case '?':  tok = Token(::QUESTION, "?"); break;
        case '\\': tok = Token(::BACKSLASH, "\\"); break;
        case '|':  tok = Token(::PIPE, "|"); break;
        default:
            // identifier
            if (isalpha(ch)) {
                string ident = readIdentifier();
                if (IDENTIFIER_MAP.find(ident) == IDENTIFIER_MAP.end())
                    return Token(::IDENTIFIER, ident);
                return Token(IDENTIFIER_MAP.at(ident), ident);
            }
            // number
            if (isdigit(ch)) {
                pair<TokenType, string> res = readNumber();
                return Token(res.first, res.second);
            }
            tok = Token(::ILLEGAL, "ILLEGAL");
    };
    advance();
    return tok;
}

string Lexer::readIdentifier() {
    uint64_t pos = curr;
    while (isalpha(ch) || isdigit(ch) || ch == '_')
        advance();
    return input.substr(pos, curr - pos);
}

pair<TokenType, string> Lexer::readNumber() {
    uint64_t pos  = curr;
    bool is_float = false;
    while (isdigit(ch) || ch == '.') {
        if (ch == '.') {
            if (is_float) {
                advance();
                return {::ILLEGAL, "ILLEGAL"};
            }
            is_float = true;
        }
        advance();
    }
    return {is_float ? ::FLOAT : ::NUMBER, input.substr(pos, curr - pos)};
}

string Lexer::readString() {
    advance(); // skip "
    uint64_t pos = curr;
    while (ch != '\"' && ch != 0)
        advance();
    return input.substr(pos, curr - pos);
}

string Lexer::readChar() {
    advance(); // skip '
    uint64_t pos = curr;
    while (ch != '\'' && ch != 0)
        advance();
    return input.substr(pos, curr - pos);
}

string Lexer::readComment() {
    uint64_t pos = curr;
    while (ch != 0 && ch != '\n' && ch != '\r')
        advance();
    return input.substr(pos, curr - pos);
}

void Lexer::advance() {
    ch   = peek >= input.length() ? 0 : input[peek];
    curr = peek++;
}

File: Python/Lexer/lexer.py
from tokens import DOUBLE_TOKEN_MAP, IDENTIFIER_MAP, Token, TokenType


class Lexer:
    curr: int
    peek: int
    input: str
    ch: str | None

    def __init__(self, input: str) -> None:
        self.input = input
        self.curr = 0
        self.peek = 0
        self.ch = None
        self.advance()

    def next_token(self) -> Token:
        tok: Token = Token(TokenType.EOF, "")

        while self.ch is not None and self.ch.isspace():
            self.advance()

        if not self.ch:
            return tok

        if self.peek < len(self.input):
            if (ds := self.input[self.curr : self.peek + 1]) == "//":
                return Token(TokenType.COMMENT, self.read_comment())
            elif (ttype := DOUBLE_TOKEN_MAP.get(ds, None)) is not None:
                self.advance()
                self.advance()
                return Token(ttype, ds)

        match self.ch:
            case "!":
                tok = Token(TokenType.EXCLAMATION, "!")
            case "@":
                tok = Token(TokenType.AT, "@")
            case "#":
                tok = Token(TokenType.HASHTAG, "#")
            case "$":
                tok = Token(TokenType.DOLLAR, "$")
            case "%":
                tok = Token(TokenType.PERCENT, "%")
            case "^":
                tok = Token(TokenType.CARET, "^")
            case "&":
                tok = Token(TokenType.AMPERSAND, "&")
            case "*":
                tok = Token(TokenType.ASTERISK, "*")
            case "(":
                tok = Token(TokenType.LPAREN, "(")
            case ")":
                tok = Token(TokenType.RPAREN, ")")
            case "-":
                tok = Token(TokenType.MINUS, "-")
            case "_":
                tok = Token(TokenType.UNDERSCORE, "_")
            case "+":
                tok = Token(TokenType.PLUS, "+")
            case "=":
                tok = Token(TokenType.ASSIGN, "=")
            case "[":
                tok = Token(TokenType.LBRACKET, "[")
            case "]":
                tok = Token(TokenType.RBRACKET, "]")
            case "{":
                tok = Token(TokenType.LBRACE, "{")
            case "}":
                tok = Token(TokenType.RBRACE, "}")
            case ";":
                tok = Token(TokenType.SEMICOLON, ";")
            case ":":
                tok = Token(TokenType.COLON, ":")
            case "'":
                tok = Token(TokenType.CHAR, self.read_char_literal())
            case '"':
                tok = Token(TokenType.STRING, self.read_string())
            case ",":
                tok = Token(TokenType.COMMA, ",")
            case ".":
                tok = Token(TokenType.PERIOD, ".")
            case "<":
                tok = Token(TokenType.LESSTHAN, "<")
            case ">":
                tok = Token(TokenType.GREATERTHAN, ">")
            case "/":
                tok = Token(TokenType.SLASH, "/")
            case "?":
                tok = Token(TokenType.QUESTION, "?")
            case "\\":
                tok = Token(TokenType.BACKSLASH, "\\")
            case "|":
                tok = Token(TokenType.PIPE, "|")
            case _:
                if self.ch.isalpha():
                    if ttype := IDENTIFIER_MAP.get((ident := self.read_identifier())):
                        return Token(ttype, ident)
                    else:
                        return Token(TokenType.IDENTIFIER, ident)
                elif self.ch.isdigit():
                    return Token(*self.read_number())
                return Token(TokenType.ILLEGAL, "ILLEGAL")

        self.advance()
        return tok

    def read_number(self) -> tuple[TokenType, str]:
        pos = self.curr
        is_float = False
        while self.ch is not None and self.ch.isdigit() or self.ch == ".":
            if self.ch == ".":
                if is_float:
                    return TokenType.ILLEGAL, "ILLEGAL"
                is_float = True
            self.advance()
        if is_float:
            return TokenType.FLOAT, self.input[pos : self.curr]
        return TokenType.NUMBER, self.input[pos : self.curr]

    def read_identifier(self) -> str:
        pos = self.curr
        while self.ch is not None and self.ch.isalnum() or self.ch == "_":
            self.advance()
        return self.input[pos : self.curr]

    def read_string(self) -> str:
        self.advance()
        pos = self.curr
        while self.ch is not None and self.ch != '"':
            self.advance()
        return self.input[pos : self.curr]

    def read_char_literal(self) -> str:
        self.advance()
        pos = self.curr
        while self.ch is not None and self.ch != "'":
            self.advance()
        return self.input[pos : self.curr]

    def read_comment(self) -> str:
        pos = self.curr
        while self.ch is not None and self.ch not in ["\n", "\r"]:
            self.advance()
        return self.input[pos : self.curr]

    def advance(self) -> None:
        if self.peek >= len(self.input):
            self.ch = None
        else:
            self.ch = self.input[self.peek]
        self.curr = self.peek
        self.peek += 1
